### ISSUE TITLE
Implement Miller columns

### ISSUE THREAD

**Issue by nkkollaw** [nkkollaw/nemo#1]

**Comment by benreu** [comment]
I just ran across this fork and am curious on what your direction is / thoughts are. I am programming Gtk3 in Python for over two years now and really would like to see Miller Columns in Nemo. Let me know if I can help.

**Comment by nkkollaw** [comment]
Hi @benreu. Unfortunately I created this fork just to post this issue on Bontysource. I'm a web developer and I wouldn't know where to start with actually writing the code :-/. I thought I'd put a few hundred dollars on Bountysource to get the ball rolling but my PayPal account wasn't working.

I think Miller columns are the most efficient way to browse a filesystem. The only file manager on Linux that has them is Pantheon Files, but it's extremely buggy and keeps crashing, which is very annoying for a file manager.

Other than that, I know Nemo has a tree view, so implementing Miller columns should take 1 month of work according to 1 person I talked to.

**Comment by benreu** [comment]
I have cloned the source and tried to build it, but meson says I am missing cinnamon-desktop. I am running Mint 18.0 so that might be the problem. 

Is there a possibility the person who thinks 1 month of work would give Miller Columns to Nemo could participate in this thread? Or is he not Gtk/desktop inclined?

I might know of another person who could help, but I never asked if he is or would be interested in Miller Columns or Nemo... let me see...

**Comment by nkkollaw** [comment]
> I have cloned the source and tried to build it, but meson says I am missing cinnamon-desktop. I am running Mint 18.0 so that might be the problem.

I'm using it with Xfce and don't have `cinnamon-desktop` installed, so I don't think it's a requirement, but I have installed via Apt. I will try building from source.

> Is there a possibility the person who thinks 1 month of work would give Miller Columns to Nemo could participate in this thread? Or is he not Gtk/desktop inclined?

Unfortunately I've asked, but he's not interested because he's building things for KDE and he doesn't seem interested in Gnome and similar.

> I might know of another person who could help, but I never asked if he is or would be interested in Miller Columns or Nemo... let me see...

Yes, I really am useless. I can help with testing of course, but very little otherwise as I have no idea how GTK works.

**Comment by nkkollaw** [comment]
@benreu: I've added installation instructions that actually work (see README). There is no dependency on `cinnamon-desktop`.

The instructions let you build .deb files for installation, I'm not sure that's ideal while developing..?

**Comment by benreu** [comment]
OK, so I found the fix online: install **libcinnamon-desktop**

Now my problem is:

Meson encountered an error in file meson.build, line 80, column 0:
Native dependency 'xapp' not found

And it doesn't seem like there is any libxapp or xapp-dev or anything pertinent. This probably will be available in Mint 19...      or 18.3 (but I only upgrade every two years). Sigh...

**Comment by benreu** [comment]
Sorry, I didn't read you comments thoroughly. Working...

**Comment by benreu** [comment]
OK, I got it compiled. Now it conflicts with my existing Nemo install. But at least I have something tangible to play with.

**Comment by nkkollaw** [comment]
Very nice!

Let me know if I can be of any help.

If needed I will take care of Bountysource.

**Comment by benreu** [comment]
If you have a moment, check out the tiny bit I got going here (feedback, etc is welcome):

https://github.com/benreu/nemo_miller_columns

A 64 bit executable here:

https://github.com/benreu/nemo_miller_columns/blob/master/Debug/src/gtk_nemo_columns

I emailed one of the Nemo devs to see if they could help me a bit incorporate my code into Nemo.

**Comment by nkkollaw** [comment]
Holy crap man, that looks awesome!

I didn't think you were working on it.

Is it far from supporting regular file operations..?

**Comment by benreu** [comment]
Yeah, lot of work to do. I am not going to reinvent the wheel for right click menus, file operations, etc... Nemo has all this as a #include, but I am stumped where to hook in my code to the main Nemo code.

The next on the list is a better preview. I haven't seen Miller Columns more than 5 years. Do you have an example or something of how other software does it? File size, type, attributes, etc...

**Comment by nkkollaw** [comment]
> Yeah, lot of work to do. I am not going to reinvent the wheel for right click menus, file operations, etc... Nemo has all this as a #include, but I am stumped where to hook in my code to the main Nemo code.

Yes, I would think that the features in Nemo can be triggered from the Miller column view, somehow. Hopefully the devs will reply to you.

> The next on the list is a better preview. I haven't seen Miller Columns more than 5 years. Do you have an example or something of how other software does it? File size, type, attributes, etc...

Finder from macOS is implemented very well. In the Linux world I only know of Pantheon Files, but the default settings are idiotic and not easy to change (for instance, single-click by default, select files via checkboxes instead of clicking on them, etc.).

There is a video here: https://www.youtube.com/watch?v=yQ7trdpY9MI, or if you have a Mac around, I can send you a VirtualBox virtual machine with a macOS installation (you need a Mac because if not you breach the license).

I don't think there are a lot more features compared to your view. The Finder in macOS has 2 status bars: one with the path, and one with info about the current selected item. There is also a preview card when you select a file.

**Comment by benreu** [comment]
Thanks a lot. I had three Macs years ago, but switched to Linux for several reasons. I just didn't remember all the little details on Finder Columns and needed a refresher. I tried Pantheon back when it was called Marlin, but I got tired of crashes and the devs weren't too active.

Thanks again.

**Comment by nkkollaw** [comment]
> Thanks a lot. I had three Macs years ago, but switched to Linux for several reasons

Yes, of course Virtual Box works on Linux, too (like you might know), but that would break macOS's license. So, if you find somebody to lend you a Mac, I can give you a file you can import into VirtualBox. Just let me know if you find a Mac and I can share it via Google Drive, but make sure you absolutely not import it into your copy of VirtualBox on Linux!

>  I tried Pantheon back when it was called Marlin, but I got tired of crashes and the devs weren't too active.

Don't worry, Pantheon Files crashes just as much. It's basically unusable, crashing in the middle of file transfers on a regular basis.

:-)

**Comment by benreu** [comment]
I'm sorry to say, but due to changes in my daily life, I won't be active on this for some time. I'll work on it as life allows me, but I can't make any promises. 

Thanks for all you've helped me including the documentation and encouragement.

So long...
